### ISSUE TITLE
Using the best match instead of the hungarian match is weird.

### ISSUE THREAD

**Issue by samuelgarcia** [SpikeInterface/spikecomparison#5]
@alejoe91 @magland @colehurwitz @mhhennig 

I am back on using the GrounfTruthComparison class.

We decided in edimburg to use the **best_match** in instead of the hungarian for the final matching to make performence with ground truth.

See https://github.com/SpikeInterface/spikecomparison/blob/master/spikecomparison/groundtruthcomparison.py#L56

This lead to very strange table **count**.. Here an example on a dataset.

Here with best match (new version)
```
tp   fn  cl  fp  num_gt  num_tested  tested_id gt_unit_id
0             0  146   0  55     146         180          4
1             0   96   0   0      96           0         -1
2             0  144   0  55     144         180          4
3           126    1   0   0     127         126          0
```

Here with hungarian match (old version):
```
tp   fn  cl  fp  num_gt  num_tested  tested_id gt_unit_id 
0             0  146   0   0     146           0         -1
1             0   96   0   0      96           0         -1
2             0  144   0   0     144           0         -1
3           126    1   0   0     127         126          0
```

As you can see with **best match** the gt unit #0 and #2  are matched to tested unit #4 with TP=0. This is totaly weird. In the second table (hugarian) it was not the case. (-1)

The main raison is the combinaison of using the **best match** + making a match with **score>0** and not score>agreement.
See https://github.com/SpikeInterface/spikecomparison/blob/master/spikecomparison/comparisontools.py#L199

I think this is a mistake. We have to change this before the release.

now spikeforest is using spiekcomparison. This could bias results. No ?

**Comment by samuelgarcia** [comment]
Note that the choice we made with matching anything with score>0 affect also the couting of redundant units.

Furthermore; for long recording.Statically all units will be matched to all other units because they wil fire at the same time at least once. And even there is no match the best match will be affected to the cell that fire the more because the chance to fire at the same time is higher.

We definitively need to change this.

**Comment by alejoe91** [comment]
Ok so we should best_match + score>min_agreement?

**Comment by alejoe91** [comment]
I agree this is an unintended behavior

**Comment by samuelgarcia** [comment]
I think I catch the bug: this is here https://github.com/SpikeInterface/spikecomparison/blob/master/spikecomparison/basecomparison.py#L79

The score label was done with hugarian (self._unit_map12).
In GT comparison it must be done with best match.

This is why this is strange.

**Comment by alejoe91** [comment]
AhAh!!! good catch!

**Comment by samuelgarcia** [comment]
OK I tried a fix on my mahcine.
This is really better but still have the weird case when the best match have TP=0.

I think I prefered the previous behavior with the hungarian in wich a best match could be only once.
So this avoid TP=0 and test_id!=-1.

**Comment by samuelgarcia** [comment]
I would also proposed to NOT match very low score agreement. I think the chance level can be rougthly estimated and so agreement near that score would not be in matched list so best match would not count then.

**Comment by magland** [comment]
A bit confusing to me. Could you explain what score is defined as? And how you could have TP=0 and score>0?

**Comment by samuelgarcia** [comment]
score is the agreement_score, so the score from all unit1 to all unit2 before the the matching.
Based on this the a match is agreement_score>0 and the best match can have a very low score_agreement.
After matching in one way, we are able to compute TP/FN/FP.

With the new choice of "best match" against "hugarian match", a unit from sorting2 can be the best match of several unit1. But only one will be the good one. The good one will have TP but other no because it is one way.

Note that the vector score label per unit is comptued only once.

Example we can have:
  * u1=0 (from gt) have best match to u2=0 (from tested) with score =0.8
  * u1=1 (from gt) have best match to u2=0 (from tested) with score = 0.04

For the first case will have many TP but the second will have 0 TP because TP of u2=0 are affected to u1=0.

I don't known if I am clear.

**Comment by magland** [comment]
My understanding is that (for the purpose of computing agreement scores)
events are not necessarily matched one-to-one. Meaning that it is allowed
for the same sorted event to be matched to more than one true event for the
purpose of computing agreement scores.

The definition of number of matches between true unit A and sorted unit B
is:

# of events of A that are within delta of an event in B.

Would you agree with this definition?

In this case it is okay for the same sorted event to be included in the
computed agreement score for more than one true unit.

**Comment by samuelgarcia** [comment]
For score agreement, this is already the case.

But for for counting TP it is not the case.

This is why some matched unit with score agreement have TP=0.

**Comment by magland** [comment]
hmmm, shouldn't we use the method I outlined for counting true positives? I
think that would be the most sensible.

**Comment by samuelgarcia** [comment]
If I anderstand correctly, you propose a counting without the knownledge of the matching.

In short, the TP counting is more or less the score_agreement except that the denominator would be the num_GT.

The main drawback, I see here is that each spike in the tesed spçketrain can be counted several time for several spiketrain in GT.

The actual implementation have change since Juanary but the main method is still the same:
agreement > matching(hugarian or best) > labeling TP/FP/TP > counting TP

I think I like it because spiek are counted only once.


Another thing to take in account is that if the recording is very long by chance all gt unit will match all tested units. the agreement score will never be zero. (it is not the case for short recording).
In that case the TP would never be 0. 
The chance level of having 2 spike within a delta could be appromiated by Bernoulli distribution. So we could remove very small agreement for teh matching.

My idea would be to implement in the same class the 2 approach but we have to be very very carrfull about the semantic. What is a TP, can it be counted 2 times.

Globally the difference between the 2 apporach will very small in general case but when we will look at spike collision resolution it will matter a lot.

**Comment by samuelgarcia** [comment]
Also see #7

**Comment by magland** [comment]
Sam, I think you have a better grasp that I do of the intricacies of how
the code actually works.

In my opinion we should define mathematically what exactly we are
computing. Then a secondary concern is whether the code is fulfilling that
requirement.

Terms like "num_matches" are not precise enough. Rather we need to say
something like

Let A be a ground truth unit and B be a sorted unit. Define num_matches
(between A and B) to be the number of events in A that are within delta
time steps of an event in B. Then the true positive rate (between A and B)
would be defined as TP = num_matches/#A, where #A is the total number of
events in A (number of ground truth events).

At least that is how I would define it. Would you agree with this
definition? If not, what is your definition?

**Comment by alejoe91** [comment]
Hi @magland 

I think that the true positive rate that you defined is actually a 'match_rate'. The 'true positive rate' is the 'matched rate' only when two spike trains are matched.

The problem with using the best match instead of the hungarian match is the following:
In labeling the ground truth spikes ([do_score_labels](https://github.com/SpikeInterface/spikecomparison/blob/master/spikecomparison/comparisontools.py#L265)) only the hungarian matched spike train is used to label TP (using [_unit_map_12](https://github.com/SpikeInterface/spikecomparison/blob/master/spikecomparison/basecomparison.py#L78), which is the hungarian match). This results in a mismatch between the matched_id (computed using the [best match](https://github.com/SpikeInterface/spikecomparison/blob/master/spikecomparison/groundtruthcomparison.py#L59) and not the hungarian match) and the number of TP, which are 0 because TP are computed using the hungarian matched unit.

I hope this clears things up. So my suggestion is to always use the hungarian match. The other units and respective match_rates can be computed using the get_redundant_units function.

I think we should discuss about this in a call. In this status, the behavior is kind of hybrid.

**Comment by magland** [comment]
Yeah I think we should have a call. In my opinion, when comparing with
ground truth, the hungarian matching should not be used at all. Here's why:
how well you match gt unit X should not depend on other ground truth units.

**Comment by magland** [comment]
(or I'll rephrase -- how well you match gt unit X should not be affected by
other ground truth units)

**Comment by alejoe91** [comment]
One way could be using best match, but discarding the duplicated matches. When scoring labels, we need a 1-to-1 gt to tested unit match for how the current implementation works now

**Comment by samuelgarcia** [comment]
Hi Jeremy.
I am not totally agree with this appraoch. Because, in my mind the score have to be affected by others.

In short for hit_rate this will not change so much unless for "chance level" hit_rate that would be zero one case and very small in other case. The miss rate will be strictly the same. But the fasle_positive will differents. Because if a score is affected by othe means that a spike can be counted sevreal time for several unit pairs and can be counted as a TP somewhere and and FP in lany more case.
I am not sur I want this.

I will write a document with pseudo code before the discusstion of today. There pro and cons in both approach. I proposal in maybe to implement the 2 possibilities with 2 differents class but with one function with a kargs option.

**Comment by samuelgarcia** [comment]
@magland @alejoe91 I have started a document directly in spikeinterface doc:
https://github.com/SpikeInterface/spikeinterface/pull/25

Could have a look before the call ?

**Comment by samuelgarcia** [comment]
Fix in #11 and #17
